File: bubus-ts/src/bridge_nats.ts
import { BaseEvent } from './base_event.js'
import { EventBus } from './event_bus.js'
import { assertOptionalDependencyAvailable, importOptionalDependency, isNodeRuntime } from './optional_deps.js'
import type { EventClass, EventHandlerFunction, EventKey, UntypedEventHandlerFunction } from './types.js'

const randomSuffix = (): string => Math.random().toString(36).slice(2, 10)

export class NATSEventBridge {
  readonly server: string
  readonly subject: string
  readonly name: string

  private readonly inbound_bus: EventBus
  private running: boolean
  private nc: any | null
  private sub_task: Promise<void> | null

  constructor(server: string, subject: string, name?: string) {
    assertOptionalDependencyAvailable('NATSEventBridge', 'nats')

    this.server = server
    this.subject = subject
    this.name = name ?? `NATSEventBridge_${randomSuffix()}`
    this.inbound_bus = new EventBus(this.name, { max_history_size: 0 })
    this.running = false
    this.nc = null
    this.sub_task = null

    this.dispatch = this.dispatch.bind(this)
    this.emit = this.emit.bind(this)
    this.on = this.on.bind(this)
  }

  on<T extends BaseEvent>(event_key: EventClass<T>, handler: EventHandlerFunction<T>): void
  on<T extends BaseEvent>(event_key: string | '*', handler: UntypedEventHandlerFunction<T>): void
  on(event_key: EventKey | '*', handler: EventHandlerFunction | UntypedEventHandlerFunction): void {
    this.ensureStarted()
    if (typeof event_key === 'string') {
      this.inbound_bus.on(event_key, handler as UntypedEventHandlerFunction<BaseEvent>)
      return
    }
    this.inbound_bus.on(event_key as EventClass<BaseEvent>, handler as EventHandlerFunction<BaseEvent>)
  }

  async dispatch<T extends BaseEvent>(event: T): Promise<void> {
    this.ensureStarted()
    if (!this.nc) await this.start()

    const payload = JSON.stringify(event.toJSON())
    this.nc.publish(this.subject, new TextEncoder().encode(payload))
  }

  async emit<T extends BaseEvent>(event: T): Promise<void> {
    return this.dispatch(event)
  }

  async start(): Promise<void> {
    if (this.running) return
    if (!isNodeRuntime()) {
      throw new Error('NATSEventBridge is only supported in Node.js runtimes')
    }

    const mod = await importOptionalDependency('NATSEventBridge', 'nats')
    const connect = mod.connect
    this.nc = await connect({ servers: this.server })
    const sub = this.nc.subscribe(this.subject)

    this.running = true
    this.sub_task = (async () => {
      for await (const msg of sub) {
        try {
          const payload = JSON.parse(new TextDecoder().decode(msg.data))
          await this.dispatchInboundPayload(payload)
        } catch {
          // Ignore malformed payloads.
        }
      }
    })()
  }

  async close(): Promise<void> {
    this.running = false
    if (this.nc) {
      await this.nc.drain()
      await this.nc.close()
      this.nc = null
    }
    await Promise.allSettled(this.sub_task ? [this.sub_task] : [])
    this.sub_task = null
    this.inbound_bus.destroy()
  }

  private ensureStarted(): void {
    if (this.running) return
    void this.start().catch((error: unknown) => {
      console.error('[bubus] NATSEventBridge failed to start', error)
    })
  }

  private async dispatchInboundPayload(payload: unknown): Promise<void> {
    const event = BaseEvent.fromJSON(payload).reset()
    this.inbound_bus.dispatch(event)
  }
}

File: bubus/bridge_nats.py
"""NATS bridge for forwarding events between runtimes.

Optional dependency: nats-py
"""

from __future__ import annotations

import asyncio
import importlib
import json
from collections.abc import Callable
from typing import Any

from uuid_extensions import uuid7str

from bubus.base_event import BaseEvent
from bubus.event_bus import EventBus, EventPatternType, in_handler_context
from bubus.helpers import QueueShutDown


class NATSEventBridge:
    def __init__(self, server: str, subject: str, *, name: str | None = None):
        self.server = server
        self.subject = subject
        self._inbound_bus = EventBus(name=name or f'NATSEventBridge_{uuid7str()[-8:]}', max_history_size=0)

        self._running = False
        self._start_task: asyncio.Task[None] | None = None
        self._start_lock = asyncio.Lock()
        self._nc: Any | None = None

    def on(self, event_pattern: EventPatternType, handler: Callable[[BaseEvent[Any]], Any]) -> None:
        self._ensure_started()
        self._inbound_bus.on(event_pattern, handler)

    async def dispatch(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        self._ensure_started()
        if self._nc is None:
            await self.start()

        payload = event.model_dump(mode='json')
        assert self._nc is not None
        await self._nc.publish(self.subject, json.dumps(payload, separators=(',', ':')).encode('utf-8'))

        if in_handler_context():
            return None
        return event

    async def emit(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        return await self.dispatch(event)

    async def start(self) -> None:
        current_task = asyncio.current_task()
        if self._start_task is not None and self._start_task is not current_task and not self._start_task.done():
            await self._start_task
            return

        if self._running:
            return

        try:
            async with self._start_lock:
                if self._running:
                    return

                nats_module = self._load_nats()
                nc = await nats_module.connect(self.server)

                async def _on_msg(msg: Any) -> None:
                    try:
                        payload = json.loads(msg.data.decode('utf-8'))
                    except Exception:
                        return
                    try:
                        await self._dispatch_inbound_payload(payload)
                    except QueueShutDown:
                        return

                try:
                    await nc.subscribe(self.subject, cb=_on_msg)
                except Exception:
                    try:
                        await nc.close()
                    except Exception:
                        pass
                    raise

                self._nc = nc
                self._running = True
        finally:
            if self._start_task is current_task:
                self._start_task = None

    async def close(self, *, clear: bool = True) -> None:
        if self._start_task is not None:
            self._start_task.cancel()
            await asyncio.gather(self._start_task, return_exceptions=True)
            self._start_task = None
        self._running = False
        if self._nc is not None:
            try:
                await self._nc.drain()
            except Exception:
                pass
            try:
                await self._nc.close()
            except Exception:
                pass
            self._nc = None
        await self._inbound_bus.stop(clear=clear)

    def _ensure_started(self) -> None:
        if self._running:
            return
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return
        if self._start_task is None or self._start_task.done():
            self._start_task = asyncio.create_task(self.start())

    async def _dispatch_inbound_payload(self, payload: Any) -> None:
        event = BaseEvent[Any].model_validate(payload).event_reset()
        self._inbound_bus.dispatch(event)

    @staticmethod
    def _load_nats() -> Any:
        try:
            return importlib.import_module('nats')
        except ModuleNotFoundError as exc:
            raise RuntimeError('NATSEventBridge requires optional dependency: pip install nats-py') from exc
